File: 강병호/9week/[11055] 가장 큰 증가하는 부분 수열.py

'''
* Author    : BYEONGHO KANG
* Date      : 2024.11.29(Fri)
* Runtime   : 88 ms
* Memory    : 31120 KB
* Algorithm : 다이나믹 프로그래밍
'''

N = int(input())
A = list(map(int, input().split()))

dp = [0] * N

def sol(N:int, A:list):
    
    for i in range(N):
        dp[i] = A[i]
    
    for i in range(1, N):
        for j in range(i):
            if A[j] < A[i]:
                dp[i] = max(dp[i], dp[j] + A[i])
    
    return max(dp)


print(sol(N, A))

File: 이정훈/9주차/[11055] 가장 큰 증가하는 부분 수열.cpp
/*
* Author    : Lee JeongHun
* Date      : 2024.11.22(Fri)
* Runtime   : 0 ms
* Memory    : 2020 KB
* Algorithm : DP
*/

#include <iostream>
#include <vector>

using namespace std;

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    // 입력
    int n;
    cin >> n;

    vector<int> arr(n);
    for(int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    // 풀이
    vector<int> dp(n);
    int sum = dp[0];
    for(int i = 0; i < n; ++i)
    {
        dp[i] = arr[i];

        for(int j = 0; j < i; ++j)
        {
            if(arr[j] < arr[i])
            {
                dp[i] = max(dp[i], arr[i] + dp[j]);
            }
        }

        sum = max(sum, dp[i]);
    }

    // 정답
    cout << sum << '\n';

    return 0;
}